### PR TITLE
Fix NPEs caused by custom head blocks from Polymer

### PR DESCRIPTION
**Pull request by eclipseisoffline**
This PR fixes a bunch of NPE exceptions similar to the one fixed by #4557, caused by custom head blocks added by Polymer.

These NPEs are caused by block states that are unexpectedly `null`. As said in #4557, I think this is caused because Geyser retrieves the block state ID from the world and tries to look this up in the JAVA_BLOCKS registry, which returns null because Geyser is looking the block up in the server side world, which will return the actual modded block and not a player head.

The block is not registered in said registry, and thus it returns null.

The method changed in `WorldManager` didn't cause an NPE on itself, however [this](https://github.com/GeyserMC/Geyser/blob/77fa37ff828cf31fc09710a5f4c9e57ab2c88944/core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/entity/player/BedrockActionTranslator.java#L333) line did, and since the method in `WorldManager` is marked as non null I thought I should implement the fix there.

### CODE REVIEW

**Review by Camotoy — APPROVED** [review]

**Comment by onebeastchris** [comment]
Couldn't we instead of doing null checks just ensure that the block state would be air?

**Comment by eclipseisoffline** [comment]
That's what I was thinking about too. I could certainly try to implement that change by simply changing all `BlockRegistries.BLOCK_STATES.get` calls into `BlockRegistries.BLOCK_STATES.getOrDefault` ones (I don't think there's that many, unless my IDE is wrong), though there may be some edge-case scenarios. I could test it tomorrow.

**Comment by eclipseisoffline** [comment]
Implemented this, `BlockState.of` is now a null-safe method that looks up a Java block state ID in the registry and defaults to air if null. This works pretty well and fixes most NPEs, except for a couple that where caused by unpacking a null block state property, in both cases I just added a default value ([here](https://github.com/GeyserMC/Geyser/blob/a433c341fc6be70c11ca50b80ab3bf808ae9dd8a/core/src/main/java/org/geysermc/geyser/entity/type/player/SkullPlayerEntity.java#L155) and [here](https://github.com/GeyserMC/Geyser/blob/a433c341fc6be70c11ca50b80ab3bf808ae9dd8a/core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java#L62)). Replaced most calls to `BlockRegistries.BLOCK_STATES.get` with `BlockState.of`.

**Review by onebeastchris — APPROVED** [review]
Looks good - just one more question :p

File: core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java [hunk 1]
```diff
@@ -58,7 +58,7 @@ public void translateTag(GeyserSession session, NbtMapBuilder bedrockNbt, NbtMap
             bedrockNbt.putFloat("Rotation", rotation * 22.5f);
         }
         bedrockNbt.putByte("SkullType", (byte) (blockState.block() instanceof SkullBlock skull ? skull.skullType().bedrockId() : 0));
-        if (blockState.getValue(Properties.POWERED)) {
+        if (blockState.getValue(Properties.POWERED, false)) {
```
[the remark this answers is not in the document]
Comment (eclipseisoffline) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java:61]: Hmm, currently this check isn't actually there - it's just assumed the block state has the powered property, and if it doesn't (for example when the block state is air because the block is modded), it causes an NPE when unpacking the Boolean that `getValue(Properties.POWERED)` returns (because it's null). I guess something like this could work:

```java
if (blockState.block() instanceof SkullBlock skull) {
    bedrockNbt.putByte("SkullType", (byte) skull.skullType().bedrockId());
    bedrockNbt.putBoolean("MouthMoving", blockState.getValue(Properties.POWERED));
} else {
    bedrockNbt.putByte("SkullType", (byte) 0);
}
```
Though I found it easier to just add a default value.

[the remark this answers is not in the document]
Comment (onebeastchris) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java:61]: True, that's easier. However, now that i think of it - did you verify that this is still an issue after <https://github.com/GeyserMC/Geyser/pull/4766/>? There's a check now before translating block entities that checks whether the block's block entity type matches the one java sends us, so ideally, this code shouldn't be called at all now for non-skull blocks now

[the remark this answers is not in the document]
Comment (eclipseisoffline) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java:61]: Didn't notice that! Will verify in a moment and then I'll get back to you

[the remark this answers is not in the document]
Comment (eclipseisoffline) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java:61]: Tested a bit, you were correct. The NPE doesn't occur with your recent PR. However, a new issue arises. Block entity data of modded blocks like the Polymer ones is now not translated at all, causing all of the custom head blocks to not have a texture and appear like a normal skeleton skull block.
A fix would be to check if the block is registered in Geyser, and if it isn't (which would mean it's modded), translate the block entity anyway, regardless of whether the type matches or not. Though this can cause issues like the aforementioned NPE.

[the remark this answers is not in the document]
Comment (onebeastchris) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java:61]: The latter would defeat the purpose of the previous PR, and it would also not be what a vanilla Java client would do (when a mismatch between block & received block entity type for said block is detected).

IMO, one possible "solution" would be a mixin from another mod the relevant world accessor methods, and ensuring that Geyser's world manager is receiving the vanilla blocks instead of the modded blocks. Alternatively, you could just register the modded blocks with Geyser's custom book system, which does support non vanilla blocks too.

[the remark this answers is not in the document]
Comment (eclipseisoffline) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java:61]: That is true, though in the case of Polymer the Java client receives the block as the actual client side block (so a player head), whereas Geyser looks the block up server side and finds the modded block (if it weren't for that, this PR wouldn't exist).
Your proposed solution sounds decent. I'll look if such a mixin is possible to implement within Polymer. For now I'll remove the default fallback from the commit, since it isn't necessary.

[the remark this answers is not in the document]
Comment (onebeastchris) [reply core/src/main/java/org/geysermc/geyser/translator/level/block/entity/SkullBlockEntityTranslator.java]: Sounds good :)
Maybe we could look into implementing a system where blocks that don't match the Java servers could be cached (aka to avoid this issue), but that would sort of defeat the purpose of having direct world access in the first place :/

unless we implement a system for caching block entities - that would have a few other applications too.
